src/Reference.js

```diff
@@ -15,34 +16,38 @@ export default class Reference {
   }
 
   constructor(key, mapFn, options = {}) {
-    validateName(key);
+    validateType(key);
     let prefix = options.contextPrefix || '$';
 
-    if (typeof key === 'function') {
-      key = '.';
-    }
-
     this.key = key.trim();
     this.prefix = prefix;
     this.isContext = this.key.indexOf(prefix) === 0;
-    this.isSelf = this.key === '.';
-
+    this.isRelativePath = isRelativePath(this.key);
     this.path = this.isContext ? this.key.slice(this.prefix.length) : this.key;
-    this._get = getter(this.path, true);
+    if (!this.isRelativePath) {
+      this._get = getter(this.path, true);
+    }
     this.map = mapFn || (value => value);
   }
   resolve() {
     return this;
   }
 
-  cast(value, { parent, context }) {
-    return this.getValue(parent, context);
+  cast(value, options) {
+    return this.getValue(options);
   }
 
-  getValue(parent, context) {
-    let isContext = this.isContext;
-    let value = this._get(isContext ? context : parent || context || {});
-    return this.map(value);
+  getValue(options) {
+    const { context, parent, path, value = {} } = options;
```

Comment:
It feels like adding another property here is not ideal. I admit I haven't had a chance to really look over everything, but i don't quite understand what `value` is here. It seems like minimally what you need to resolve a negative path is the entire nested object, the current `path` to and then the relative path from there. what about changing `parent` to an array of ancestors? and using that to resolve everything?